snuba/data_delete.py

```diff
@@ -0,0 +1,108 @@
+from collections import defaultdict
+from enum import Enum
+from time import sleep
+from typing import Any, List, Mapping, Sequence
+
+import structlog
+from structlog.contextvars import bind_contextvars, unbind_contextvars
+
+from snuba.clickhouse.native import ClickhousePool, ClickhouseResult
+from snuba.clusters.cluster import (
+    ClickhouseClientSettings,
+    ClickhouseCluster,
+    ClickhouseNode,
+    connection_cache,
+)
+
+logger = structlog.get_logger().bind(module=__name__)
+
+
+class DeleteResult(Enum):
+    IN_PROGRESS = 0
+    FAILED = 1
+    SUCCEEDED = 2
+
+
+def order_cluster_nodes(
+    nodes: Sequence[ClickhouseNode],
+) -> Mapping[int, Sequence[ClickhouseNode]]:
+    """
+    Given a series of unordered ClickHouseNode[s], arrange them by shard
+    """
+    ordered_nodes: Mapping[int, List[ClickhouseNode]] = defaultdict(list)
+    for node in nodes:
+        assert (
+            node.shard is not None
+        ), f"cannot proceed with automatic delete, node {node} has empty shard"
+        ordered_nodes[node.shard].append(node)
+    return ordered_nodes
+
+
+def get_working_replica_node(
+    nodes: Sequence[ClickhouseNode], user: str, password: str, database: str
+) -> ClickhousePool:
+    # TODO actually fail to getting another node if we can't load this connection
+    return connection_cache.get_node_connection(
+        ClickhouseClientSettings.DATA_DELETE, nodes[0], user, password, database
+    )
+
+
+def execute_delete_command_on_replica(
+    connection: ClickhousePool,
+    query: str,
+    params: Mapping[str, Any],
+    dry_run: bool = False,
+) -> ClickhouseResult:
+    logger.info("data_delete.pre_delete", interpolated_query={query % params})
+    if dry_run:
+        return ClickhouseResult()
+    else:
+        return connection.execute(query, params, retryable=False)
```

Comment:
Are you planning to execute the mutation per part or per partition ? And who is going to restrict the query to a specific partition.

Reply:
I thought that this had to be controlled by creating a safe input template and parameters. Are there actually query/DDL execution settings in ClickHouse that will let you manually limit the scope of a single query to a part or partition? 

Alternatively, were you hoping to programmatically verify that a delete will not span partitions by parsing the delete query conditions and comparing it to the table schema? That seems complicated to me, but I'd be fine with investigating.

Reply:
Ideally, this would be run per partition, not parts. Partitions are easier to keep track of. Can you check if our clickhouse version supports this https://clickhouse.com/docs/en/sql-reference/statements/alter/partition/#delete-in-partition. If it does, then you can get all the active partitions and apply the command in sequence to each one of them. You probably won't even need this condition then https://github.com/getsentry/snuba/pull/3460/files#diff-e6a3c68f04d4785448cb8190f9633b0e7109857bdbcaccc8898b96603bd1f89dR14.

Reply:
I'll take a look at whether our ClickHouse versions support `DELETE IN PARTITION`